monai/apps/pathology/transforms/post/array.py

```diff
@@ -0,0 +1,307 @@
+        Args:
+            image: image where the lowest value points are labeled first. Shape must be [1, H, W].
+            mask: the same shape as image. Only points at which mask == True will be labeled.
+            markers: The desired number of markers, or an array marking the basins with the values to be assigned
+                in the label matrix. Zero means not a marker. If None (no markers given), the local minima of the
+                image are used as markers. Shape must be [1, H, W].
+        """
+
+        image = convert_to_numpy(image)
+        markers = convert_to_numpy(markers)
+        mask = convert_to_numpy(mask)
+
+        if image.shape[0] != 1 or image.ndim != 3:
+            raise ValueError(f"Input image should be with size of [1, H, W], but got {image.shape}")
+        if mask.shape[0] != 1 or mask.ndim != 3:
+            raise ValueError(f"Input mask should be with size of [1, H, W], but got {mask.shape}")
+        if markers is not None and (markers.shape[0] != 1 or markers.ndim != 3):
+            raise ValueError(f"Input markers should be with size of [1, H, W], but got {markers.shape}")
+
+        instance_seg = watershed(image, markers=markers, mask=mask, connectivity=self.connectivity)
+
+        return convert_to_dst_type(instance_seg, image, dtype=np.uint8)[0]
+
+
+class GenerateMask(Transform):
+    """
+    generate mask used in `watershed`. Only points at which mask == True will be labeled.
+
+    Args:
+        softmax: if True, apply a softmax function to the prediction.
+        sigmoid: if True, apply a sigmoid function to the prediction.
+        threshold: if not None, threshold the float values to int number 0 or 1 with specified theashold.
+        remove_small_objects: whether need to remove some objects in the marker. Defaults to True.
+        min_size: objects smaller than this size are removed if `remove_small_objects` is True. Defaults to 10.
+
+    Raises:
+        ValueError: when the `prob_map` shape is not [C, H, W].
+
+    """
+
+    backend = [TransformBackends.NUMPY]
+
+    def __init__(
+        self,
+        softmax: bool = True,
+        sigmoid: bool = False,
+        threshold: Optional[float] = None,
+        remove_small_objects: bool = True,
+        min_size: int = 10,
+    ) -> None:
+        if sigmoid and threshold is None:
+            raise ValueError("Threshold is needed when using sigmoid activation.")
+
+        self.activations = Activations(sigmoid=sigmoid, softmax=softmax)
+        self.asdiscrete = AsDiscrete(threshold=threshold, argmax=softmax)
+        if remove_small_objects:
+            self.remove_small_objects = RemoveSmallObjects(min_size=min_size)
+        else:
+            self.remove_small_objects = None  # type: ignore
+
+    def __call__(self, prob_map: NdarrayOrTensor) -> NdarrayOrTensor:
+        """
+        Args:
+            prob_map: probability map of segmentation, shape must be [C, H, W]
+        """
+        if len(prob_map.shape) != 3:
```

Comment:
Hi @wyli, you are right, `GenerateMask` can support cxhxwxd shape as well, but I'm not sure whether we need to support that since I have declared in this place that the generated mask is used for `watershed`. What do you think?
https://github.com/Project-MONAI/MONAI/blob/9d75e66e4deb027542e214878b143695c4dfac06/monai/apps/pathology/transforms/post/array.py#L85

Reply:
watershed also supports 3d

Reply:
oh, l see, I will update it, thanks!